src/env.cc

```diff
@@ -1100,6 +1108,51 @@ void Environment::BuildEmbedderGraph(Isolate* isolate,
   Environment* env = static_cast<Environment*>(data);
   tracker.Track(env);
 }
+size_t Environment::NearHeapLimitCallback(void* data,
+                                          size_t current_heap_limit,
+                                          size_t initial_heap_limit) {
+  Environment* env = static_cast<Environment*>(data);
+
+  Debug(env,
+        DebugCategory::DIAGNOSTICS,
+        "Invoked NearHeapLimitCallback, processing=%d, "
+        "current_limit=%" PRIu64 ", "
+        "initial_limit=%" PRIu64 "\n",
+        env->is_processing_heap_limit_callback_,
+        static_cast<uint64_t>(current_heap_limit),
+        static_cast<uint64_t>(initial_heap_limit));
+
+  // Since generating heap snapshot triggers GC, avoid making this
+  // a very eager GC scheduler.
+  if (current_heap_limit > initial_heap_limit) {
+    return initial_heap_limit;
+  }
+  if (env->is_processing_heap_limit_callback_) {
+    return initial_heap_limit;
+  }
+
+  env->is_processing_heap_limit_callback_ = true;
+  DiagnosticFilename name(env, "Heap", "heapsnapshot");
+  heap::WriteSnapshot(env->isolate(), *name);
+  env->heap_limit_snapshot_taken_ += 1;
+  if (env->heap_limit_snapshot_taken_ ==
+      env->options_->heap_snapshot_near_heap_limit) {
+    env->isolate()->RemoveNearHeapLimitCallback(NearHeapLimitCallback,
+                                                initial_heap_limit);
+  }
+  Debug(env, DebugCategory::DIAGNOSTICS, "Wrote snapshot to %s\n", *name);
+  env->isolate()->AutomaticallyRestoreInitialHeapLimit(1.0);
+
+  env->is_processing_heap_limit_callback_ = false;
+
+  // // Increase the heap size by 2
+  // const size_t kSnapshotHeapSizeFactor = 2;
+  // size_t max_base =
+  //     std::numeric_limits<size_t>::max() / kSnapshotHeapSizeFactor;
+  // size_t new_limit =
+  //     std::min(max_base, initial_heap_limit) * kSnapshotHeapSizeFactor;
+  return initial_heap_limit + 1;
```

Comment:
It's not yet clear to me why this is already enough for testing various kinds of heaps locally, despite that I had been thinking about the same thing as @bnoordhuis said in https://github.com/nodejs/node/issues/27552 But I do see that in the snapshot generator implementation it seems V8 is mostly using memory outside of the managed heap - so, for applications running in environment with more memory than the amount available to V8's managed heap, this should be alright. I am not sure what would happen if the amount of memory available to V8's heap is also the roughly same amount of memory available to the entire process (but then Node.js always need a chunk of that for its own thing, and whether V8 takes a larger chunk or Node.js takes a larger chunk really depends on what the application does)

Reply:
>  I am not sure what would happen if the amount of memory available to V8's heap is also the roughly same amount of memory available to the entire process

I believe thats the case for workload within containers, which a lot of people is using to run in production. I think the flag should have a big warning inside the documentation for those cases

Reply:
We should be able to test this with `docker run -m <memory>` to limit the amount of memory available for the process.

Reply:
Discussions in [v8-users](https://groups.google.com/forum/#!topic/v8-users/72NU6OeC3kU) suggest that we could increase the heap limit by the max young generation size to avoid crashing during the GC triggered by snapshot generation (since it may promote the objects from the young generation into old generation, increasing the heap usage).

I guess we could do some clever tricks, for example, when we initially configure the ResourceConstraints (where we already try to guess the amount of memory available to the process), we leave some leeway for heap snapshots if the option is set. And before we take the snapshots, we can use `GetHeapObjectStatisticsAtLastGC()` etc. to estimate the native memory overhead of the snapshot generation, and give up if it seems risky. We can try experimenting with the formula while this is, um, experimental.

Reply:
I used the max young generation size as configured in the ResourceConstraints to compute the new limit there is a caveat. I don't think `GetHeapObjectStatisticsAtLastGC()` would be useful here after all considering there are 1000+ types of objects that we'll need to iterate through to collect the total object count. So right now I am only including the max young generation size into the potential overhead. When taking the snapshot, if the memory available to the process isn't enough to cover the overhead, it will just give up taking the snapshot to be safe.